frontend/src/components/Tree/Tree.jsx

```diff
@@ -0,0 +1,138 @@
+import React, { useState, useRef, useEffect } from "react";
+import styles from "./Tree.module.css";
+import { select, hierarchy, tree, linkHorizontal } from "d3";
+import ResizeObserver from "resize-observer-polyfill";
+import "./Tree.module.css";
+
+const useResizeObserver = (ref) => {
+  const [dimensions, setDimensions] = useState(null);
+  useEffect(() => {
+    const observeTarget = ref.current;
+    const resizeObserver = new ResizeObserver((entries) => {
+      entries.forEach((entry) => {
+        setDimensions(entry.contentRect);
+      });
+    });
+    resizeObserver.observe(observeTarget);
+    return () => {
+      resizeObserver.unobserve(observeTarget);
+    };
+  }, [ref]);
+  return dimensions;
+};
+
+function usePrevious(value) {
+  const ref = useRef();
+  useEffect(() => {
+    ref.current = value;
+  });
+  return ref.current;
+}
+
+export function Tree({ jsonData, onNodeClick }) {
+  const [currentView, setCurrentView] = useState({});
+  const svgRef = useRef();
+  const wrapperRef = useRef();
+  const dimensions = useResizeObserver(wrapperRef);
+
+  // We save data to see if it changed
+  const previouslyRenderedData = usePrevious(currentView);
+
+  useEffect(() => {
+    setCurrentView(jsonData);
+  }, [jsonData]);
+
+  // Will be called initially and on every data change
+  useEffect(() => {
+    const svg = select(svgRef.current);
```

Comment:
As a follow-up, do you think it is good to have the drawing of the tree as part of the actual returned component rather than as an effect which is run afterwards? Because I cannot think of a situation in which we would not re render the tree itself

Reply:
For b), I'm pretty sure useEffect is exactly what we want to use because if we put this code outside the useEffect, then it would be called at every re-render, instead of just when one of the dependencies changes.

Reply:
Abstracted the tree updating and animating to two separate functions